### ISSUE TITLE
'descend' argument in 'walk' method in twisted.python.filepath._PathHelper class doesn't work well

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3911]
|<img alt="Markon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Markon reported|
|-|-|
|Trac ID|trac#3911|
|Type|defect|
|Created|2009-07-12 10:32:22Z|
|Branch|https://github.com/twisted/twisted/tree/filepath-symlink-cycle-3911|

The 'walk' method in the class _PathHelper, located in twisted/python/filepath.py doesn't work according to the 'descend' argument passed.
This is an example:

```
>>> import twisted.python.filepath as FP
>>> filep = FP.FilePath("/home/marco/Desktop")
>>> for x in filep.walk(lambda path: not path.islink()):
...     print x
...
FilePath('/home/marco/Desktop')
FilePath('/home/marco/Desktop/amarok.desktop')
[...]
FilePath('/home/marco/Desktop/MyFolder')
FilePath('/home/marco/Desktop/MyFolder/myfile.txt')
[...]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.5/site-packages/twisted/python/filepath.py", line 214, in walk
    raise LinkError("Cycle in file graph.")
twisted.python.filepath.LinkError: Cycle in file graph.

```

In the Desktop directory there's a symlink "Risorse", which points to "/home/marco". So it creates a cycle, ok. But I've specified the descend argument and it shouldn't follow the link.

However, I've also provided a patch, and if I use it, the argument 'descend' is read fine:

```
>>> import filepath as FP
>>> filep = FP.FilePath("/home/marco/Desktop")
>>> for x in filep.walk(lambda path: not path.islink()):
...     print x
...
FilePath('/home/marco/Desktop')
FilePath('/home/marco/Desktop/amarok.desktop')
[...]
FilePath('/home/marco/Desktop/MyFolder')
FilePath('/home/marco/Desktop/MyFolder/myfile.txt')
[...]
FilePath('/home/marco/Desktop/Risorse') <<<---- this is the cyclic symlink
[... other files ...]

```

I should write the tests, but in the meantime I attach the patch I've written.

Attachments:

* [filepath.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afa/afae31681987fca8fe073c04782985c77c0b22ad/6c8516bda9e085e109aecbceb95d156799596efd.patch) (1267 bytes) - added by Markon on 2009-07-12 10:32:58Z - This is the patch provided for '_PathHelper.walk'
* [filepath_test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afa/afae31681987fca8fe073c04782985c77c0b22ad/f26f628935ed7ef6e1fb7bc9b4358f641162c29f.patch) (1055 bytes) - added by Markon on 2009-07-12 13:20:00Z - This is the new test added to FilePathTestCase
* [filepath_test.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afa/afae31681987fca8fe073c04782985c77c0b22ad/bff06e6f4df33901c584ab2125319e52b50b4bf2.patch) (1055 bytes) - added by Markon on 2009-07-12 13:20:41Z - This is the new test added to FilePathTestCase

<details><summary>Searchable metadata</summary>

```
trac-id__3911 3911
type__defect defect
reporter__Markon Markon
priority__high high
milestone__ 
branch__branches_filepath_symlink_cycle_3911 branches/filepath-symlink-cycle-3911
branch_author__Markon Markon
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1247394742000000 1247394742000000
changetime__1254341516000000 1254341516000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Markon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|Markon commented|
|-|-|

I've added a little test with some cyclical symlinks in FilePathTestCase (twisted/test/test_paths.py):

```
    def test_walkObeysDescendWithCyclicalSymlinks(self):
        """
        Verify that, after making a path with cyclical symlinks,
        when the supplied C{descend} predicate returns C{False},
        the target is not traversed, as if it was a simple symlink.
        """
        self.createLinks()
        # we create cyclical symlinks
        self.symlink(self.path.child("sub1").path,
                     self.path.child("sub1").child("sub1.loopylink").path)
        def noSymLinks(path):
            return not path.islink()
        def iterateOverPath():
            return [foo.path for foo in self.path.walk(descend=noSymLinks)]
        self.assertTrue(iterateOverPath())
```

I'm sorry if this test is not pretty good, but it's the first time I write unittests for "official" works :)
However, if you use the "old" version of twisted.python.filepath.FilePath (without the patch), it Fail, since it would raise an exception: filepath.LinkErrror.
Infact:

```
[...]
test_walkObeysDescendWithCyclicalSymlinks ...                       [ERROR]
[ERROR]: test_paths.FilePathTestCase.test_walkObeysDescendWithCyclicalSymlinks

Traceback (most recent call last):
  File "test_paths.py", line 366, in test_walkObeysDescendWithCyclicalSymlinks
    self.assertTrue(iterateOverPath())
  File "test_paths.py", line 365, in iterateOverPath
    return [foo.path for foo in self.path.walk(descend=noSymLinks)]
  File "/usr/lib/python2.5/site-packages/twisted/python/filepath.py", line 211, in walk
    for subc in c.walk(descend):
  File "/usr/lib/python2.5/site-packages/twisted/python/filepath.py", line 214, in walk
    raise LinkError("Cycle in file graph.")
twisted.python.filepath.LinkError: Cycle in file graph.
```

Instead, if you change the code to use the new "walk" method, it works well, without fails.

```
test_walkObeysDescendWithCyclicalSymlinks ...                          [OK]
```


I attach the "patch" to twisted/test/test_paths.py

**Comment by twisted-trac** [comment]
|<img alt="Markon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Markon removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [27154]) Branching to 'filepath-symlink-cycle-3911'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [27155]) Apply (manually) filepath.patch and filepath_test.2.patch from Markon

refs [#3911](https://github.com/twisted/twisted/issues/3911)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

It looks reasonable, and doesn't break any other tests. Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set status to `closed`|
|-|-|

(In [27362]) Merge filepath-symlink-cycle-3911

Author: Markon
Reviewer: therve
Fixes [#3911](https://github.com/twisted/twisted/issues/3911)

Fix a bug with the descend parameter of the walk method, calling it earlier
to be able to catch cycles sooner.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
